### ISSUE TITLE
Connection string with client credentials doesn't work

### ISSUE THREAD

**Issue by Yiogi1** [MscrmTools/MscrmTools.Xrm.Connection#120]
Every time I try to create a new connection using AuthType=ClientSecret I get the error message below:
System.NullReferenceException: Object reference not set to an instance of an object.
   at McTools.Xrm.Connection.ConnectionDetail.get_UseOnline()

This was actually working in older toolbox versions as I have set up some on my old laptop. I have also tried to simply use the new method of clientid and clientsecret but then I get the following error (not that the same appid and key work fine in my C# app with both connection string and WebAPI OAuth):
ERROR REQUESTING Token FROM THE Authentication context - General ADAL Error
No redirectUri was configured. ADAL does not provide any defaults.Unable to connect to CRM: No redirectUri was configured. ADAL does not provide any defaults.
No redirectUri was configured. ADAL does not provide any defaults.Unable to Login to Dynamics CRM

---
- Deployment: Online
- DB Version: 9.1.0.16541
- XTB Version: 1.2020.4.38

**Comment by MscrmTools** [comment]
Can you test the following connection files? just replace them in your XrmToolBox installation folder

[McTools.Xrm.Connection.WinForms.zip](https://github.com/MscrmTools/MscrmTools.Xrm.Connection/files/4605688/McTools.Xrm.Connection.WinForms.zip)

**Comment by MscrmTools** [comment]
I should have precise: the updated files should allow to store client secret (which was never the case before) and ask for client secret if not stored.

I’m not sure about the nullReferenceExceptio n since I did not reproduce the error. Could you precise what connection string tag you used?

**Comment by MscrmTools** [comment]
Ping @Yiogi1 in case you receive notification only if tagged

**Comment by Paca** [comment]
> Can you test the following connection files? just replace them in your XrmToolBox installation folder
> 
> [McTools.Xrm.Connection.WinForms.zip](https://github.com/MscrmTools/MscrmTools.Xrm.Connection/files/4605688/McTools.Xrm.Connection.WinForms.zip)

I was having the same issue, after applying your change it's working again for me ;)
Thanks for the fix.

**Comment by MscrmTools** [comment]
Ok, I will merge and release this update soon

**Comment by Yiogi1** [comment]
Sorry for the delay in answering, was off for a few days. Connection string works now with this fix. The issue with the client id and key method still has the same problem but I'm perfectly happy the first method works. Don't need both. thanks for solving this

**Comment by MscrmTools** [comment]
With the client id and secret, I updated the Ui so that the password can be stored and if it’s not the code now ask for the password. Doesn’t  it fix your issue?

**Comment by Yiogi1** [comment]
I think I need to provide more context into this. I posted two different issues. The first one is with the "connection string" authentication method. This one works perfectly now, been using it for a week so I'm happy with it.

The second issue is when I try to use the "Client Id / Secret" method, this is the second error message I posted above. The ADAL error and redirect URI issue is still there when i try to use this method. I'm not sure if  you want to dig into this further. It's fine by me if you don't, I can use connection string and get my work done.
